### ISSUE TITLE
How can I define awaitable Mock or override populating function to set Future() attributes?

### ISSUE THREAD

**Issue by HoshiYamazaki** [FactoryBoy/factory_boy#702]
So we got:
https://stackoverflow.com/questions/51394411/python-object-magicmock-cant-be-used-in-await-expression

Code:
```
async def test_that_mock_can_be_awaited():
    mock = MagicMock(return_value=Future())
    mock.return_value.set_result(123)
    result = await mock()
    assert result == 123
```

I want to do:
```
class ObjectFactory(factory.Factory):

    class Meta:
        abstract = True
        model = Mock(return_value=Future())

    @classmethod
    def _create(cls, model_class, *args, **kwargs):
        async def create_coro(*args, **kwargs):
            return await model_class(*args, **kwargs)

        return create_coro(*args, **kwargs)
```

But I must populate it using `mock.return_value.set_result(123)` logic. Can I override populating function to do it?

**Comment by rbarrois** [comment]
Hi!

Just a quick note: this type of usage question is more suited for StackOverflow, where other users would be more likely to search.

Regarding your question, let's start with the API you'd expect; I understand you want to be able to call (in your code):

```python
async def test_mock_query(self):
    query = MockQueryFactory(result=123)
    result = await query()
    assert result == 123
```

For this, I think this definition should work:

```python
class MockQueryFactory(factory.Factory):
    class Meta:
        model = mock.MagicMock

    class Params:
        result = 0

    return_value = factory.LazyAttribute(lambda o: Future())

    set_forced_result__value = factory.SelfAttribute('..result')

    @factory.post_generation
    def set_forced_result(instance, create, extracted, value=None, **kwargs):
        instance.return_value.set_result(value)
```

Can you confirm what it does for you?

**Comment by HoshiYamazaki** [comment]
> Hi!
> 
> Just a quick note: this type of usage question is more suited for StackOverflow, where other users would be more likely to search.
> 
> Regarding your question, let's start with the API you'd expect; I understand you want to be able to call (in your code):
> 
> ```python
> async def test_mock_query(self):
>     query = MockQueryFactory(result=123)
>     result = await query()
>     assert result == 123
> ```
> 
> For this, I think this definition should work:
> 
> ```python
> class MockQueryFactory(factory.Factory):
>     class Meta:
>         model = mock.MagicMock
> 
>     class Params:
>         result = 0
> 
>     return_value = factory.LazyAttribute(lambda o: Future())
> 
>     set_forced_result__value = factory.SelfAttribute('..result')
> 
>     @factory.post_generation
>     def set_forced_result(instance, create, extracted, value=None, **kwargs):
>         instance.return_value.set_result(value)
> ```
> 
> Can you confirm what it does for you?

Yes, that works nicely. What I really want is to create object using Factory, where I could call abstract object method and get result.

So we got:
```python
class TelethonChatFactory(ObjectFactory):
    """Factory for Telethon Chat."""

    id = factory.Sequence(lambda n: n)
    title = factory.fuzzy.FuzzyText()
```

And:
```python
class TestChat(TestCase):
    """Test Chat."""

    async def setUp(self):
        """Prepare test for TestChat."""
        self.chat = TelethonChatFactory()

    @patch('bot.main.send_chat_message')
    async def test_chat(self, func):
        # lets assume get_chat() is telethon event function
        # we need to use get_chat() to get ID of chat, and then send message using this id
        # also, get_chat is async function
        func.event.get_chat = self.chat
        await send_chat_message.__wrapped__(func.event)
```

Let assume send_chat_message looks like that (I use Telethon anology / note - this is not the exact Telethon logic):
```python
@client.on(events.NewMessage())
async def send_chat_message(event):
    chat = await event.get_chat()
    await event.respond(chat.id, 'Something something')
```

So, I will have troubles getting .id attribute because now it's not possible (it is, if I will not do await, but I need that, since I'm testing my Telethon bot and without that it's not possible to write Telethon logic).

Also it's possible to shorten FactoryBoy logic for async testing? Seems I am somehow forced to write bunch of not clear for first eye factories to get that working. I rather would choose way that is clean and works of box (I don't care if logic is hidden under the `factory.Factory` methods).
